File: myExpenses/src/main/java/org/totschnig/myexpenses/ui/ButtonWithDialog.java
package org.totschnig.myexpenses.ui;

import android.app.Dialog;
import android.content.Context;
import android.content.ContextWrapper;
import android.os.Parcelable;
import android.util.AttributeSet;
import android.view.View;

import com.google.android.material.button.MaterialButton;

import androidx.annotation.NonNull;
import butterknife.ButterKnife;
import icepick.Icepick;

public abstract class ButtonWithDialog extends MaterialButton {

  public ButtonWithDialog(Context context, AttributeSet attrs) {
    super(context, attrs);
    ButterKnife.bind(this);
    setOnClickListener(v -> showDialog());
  }

  public void showDialog() {
    getHost().hideKeyBoardAndShowDialog(getId());
  }

  @NonNull
  @Override public Parcelable onSaveInstanceState() {
    return Icepick.saveInstanceState(this, super.onSaveInstanceState());
  }

  @Override public void onRestoreInstanceState(Parcelable state) {
    super.onRestoreInstanceState(Icepick.restoreInstanceState(this, state));
    update();
  }

  protected abstract void update();

  protected Host getHost() {
    Context context = getContext();
    while (context instanceof android.content.ContextWrapper) {
      if (context instanceof Host) {
        return (Host)context;
      }
      context = ((ContextWrapper)context).getBaseContext();
    }
    throw new IllegalStateException("Host context does not implement interface");
  }

  public interface Host {
    void hideKeyBoardAndShowDialog(int id);
    void onValueSet(View v);
  }

  public abstract Dialog onCreateDialog();
}

File: myExpenses/src/main/java/org/totschnig/myexpenses/ui/ButtonWithDialog.kt
package org.totschnig.myexpenses.ui

import android.content.Context
import android.content.ContextWrapper
import android.os.Parcelable
import android.util.AttributeSet
import android.view.View
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.FragmentActivity
import androidx.fragment.app.FragmentManager
import com.evernote.android.state.State
import com.evernote.android.state.StateSaver
import com.google.android.material.button.MaterialButton
import org.totschnig.myexpenses.util.crashreporting.CrashHandler
import org.totschnig.myexpenses.util.ui.getActivity

abstract class ButtonWithDialog<T: DialogFragment> @JvmOverloads constructor(
        context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : MaterialButton(context, attrs, defStyleAttr) {

    abstract val fragmentTag: String

    @State
    var dialogShown = false

    private fun showDialog() {
        val picker = buildDialog()
        attachListener(picker)
        picker.show((context.getActivity() as FragmentActivity).supportFragmentManager, fragmentTag)
        dialogShown = true
    }

    abstract fun buildDialog(): T

    override fun onSaveInstanceState(): Parcelable {
        return StateSaver.saveInstanceState(this, super.onSaveInstanceState())
    }

    override fun onRestoreInstanceState(state: Parcelable?) {
        super.onRestoreInstanceState(StateSaver.restoreInstanceState(this, state))
        update()
        if (dialogShown) {
            reAttachListener((context.getActivity() as FragmentActivity).supportFragmentManager)
        }
    }

    abstract fun update()

    abstract fun attachListener(dialogFragment: T)

    private fun reAttachListener(supportFragmentManager: FragmentManager) {
        (supportFragmentManager.findFragmentByTag(fragmentTag) as? T)?.apply {
            attachListener(this)
        } ?: kotlin.run {
            CrashHandler.report(Exception("reAttachListener failed"))
        }
    }

    protected val host: Host
        get() {
            var context = context
            while (context is ContextWrapper) {
                if (context is Host) {
                    return context
                }
                context = context.baseContext
            }
            throw IllegalStateException("Host context does not implement interface")
        }

    interface Host {
        fun onValueSet(view: View)
    }

    open fun onClick() {
        showDialog()
    }

    final override fun setOnClickListener(l: OnClickListener?) {
        super.setOnClickListener(l)
    }

    init {
        setOnClickListener { onClick() }
    }
}